### DISCUSSION TITLE
指定工具链构建怎么复用已经下载的库？

### DISCUSSION THREAD

**Discussion by fonqL** [xmake-io/xmake#2483] [Q&A]
**Xmake 2.67, Win10**

一开始直接用xrepo install boost下载boost，但发现只能用xmake -w 默认构建，如果指定了工具链就会提示还要下载boost。还有一个modify选项不知道是想修改啥，不过也没用

```shell
PS E:\CLionProjects\untitled> xmake -w
checking for Microsoft Visual Studio (x64) version ... 2022
[ 20%]: ccache compiling.release src\main.cpp
[100%]: build ok!

PS E:\CLionProjects\untitled> xmake f --toolchain=mingw
checking for platform ... windows
checking for architecture ... x64
checking for mingw directory ... D:\MinGW
note: install or modify (m) these packages (pass -y to skip confirm)?
in xmake-repo:
  -> boost 1.78.0 [toolchains:"mingw", vs_runtime:"MT"]
please input: y (y/n/m)
m
no more packages!
note: install or modify (m) these packages (pass -y to skip confirm)?
in xmake-repo:
  -> boost 1.78.0 [toolchains:"mingw", vs_runtime:"MT"]
please input: y (y/n/m)
n
error: packages(boost): must be installed!

PS E:\CLionProjects\untitled> xmake f --toolchain=msvc
checking for platform ... windows
checking for architecture ... x64
checking for Microsoft Visual Studio (x64) version ... 2022
note: install or modify (m) these packages (pass -y to skip confirm)?
in xmake-repo:
  -> boost 1.78.0 [toolchains:"msvc", vs_runtime:"MT"]
please input: y (y/n/m)
m
no more packages!
note: install or modify (m) these packages (pass -y to skip confirm)?
in xmake-repo:
  -> boost 1.78.0 [toolchains:"msvc", vs_runtime:"MT"]
please input: y (y/n/m)
n
error: packages(boost): must be installed!
```

xmake全局配置：
```shell
PS E:\CLionProjects\untitled> xmake g
configure
{
    theme = powershell
    pkg_installdir = E:/cpppkg
    network = public
    pkg_searchdirs = E:/cpppkg
    proxy_pac = pac.lua
    quiet = true
    cachedir = E:/xpkg_cache
}
```

然后就尝试用xrepo info/fetch --config 设置了vs_runtime和toolchain，但果然没解决问题。（其实我也不知道这个config在设置什么？希望文档完善一下

所以，我要怎么才能在指定了工具链参数后用上已经下载的boost库？其他问题还有为什么默认构建就可以？库和工具链有关系吗？构建过程中的modify选项是什么意思？xrepo info/fetch是展示信息，但是--config选项是是在配置什么？以及是不是不应该使用xrepo？

**Comment by waruqi** [comment]
> 一开始直接用xrepo install boost下载boost，但发现只能用xmake -w 默认构建，如果指定了工具链就会提示还要下载boost。还有一个modify选项不知道是想修改啥，不过也没用

正常，就是要重新下载编译，每个编译工具链生成的库 都是二进制不兼容的，还依赖编译器提供的一些 runtime，你不重新下载装，复用原有的库，最后 链接也过不了。。

就像 mingw 生成的静态库 msvc 去链接时不兼容，库格式不对。需要额外的格式转换才行。。

其他编译器也是如此。。所以不要想着怎么复用了，即使有些编译工具链生成的二进制完全兼容，但是目前 xmake 的包管理已经足够复杂，只能简单的对 toolchain 名，进行兼容性区分，是无法更进一步细粒度的去识别每个编译器之间兼容到什么地步，暂时也没精力去整这个。

**Reply by fonqL** [reply]
感谢回答！
但还有一些疑惑还请见谅。。
但是默认构建的工具链与我手动指定的工具链都一样（如上的xmake -w 和 xmake f --toolchain=msvc）时为什么也会提示重新下载呢？
还有上面提到的一些问题希望也能得到大佬的解答！
> 构建过程中的modify选项是什么意思？xrepo info/fetch是展示信息，但是--config选项是是在配置什么？

**Reply by waruqi** [reply]
> 但是默认构建的工具链与我手动指定的工具链都一样（如上的xmake -w 和 xmake f --toolchain=msvc）时为什么也会提示重新下载呢？

目前仅仅只是简单的获取 toolchain 字符串算的 hash。。默认不设置工具链，走的是 工具链探测模式，有可能完全相同，也有可能部分差异的，比如 你环境变量里改了 $CC $CFLAGS 什么的 就有差异了。。所以目前是完全区分开来的。。

即使完全一样，但是处理细分 逻辑会非常复杂的。。所以目前只是简单快速的区分 --toolchains 工具链
